tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/AutoClientIDShardClusterTest.java

```diff
@@ -0,0 +1,343 @@
+      sub1.consumerDone.set(true);
+
+      stopServers(0, 1);
+   }
+
+   @Test
+   public void testWithConsistentHashClientIDModTwo() throws Exception {
+      setupServers();
+
+      addBalancerWithClientIdConsistentHashMod(2);
+
+      startServers(0, 1);
+
+      // two bouncy durable consumers
+      DurableSub sub0 = new DurableSub("0");
+      DurableSub sub1 = new DurableSub("1");
+
+      new Thread(sub0).start();
+      new Thread(sub1).start();
+
+      // waiting for registration before production to give bridges a chance
+      assertTrue(sub0.registered.await(5, TimeUnit.SECONDS));
+      assertTrue(sub1.registered.await(5, TimeUnit.SECONDS));
+
+
+      // wait for bindings!
+      assertTrue(Wait.waitFor(() -> 1 == gatherBindingsOnT(0)[0], 20000, 2000));
+
+      assertTrue(Wait.waitFor(() -> 1 == gatherBindingsOnT(1)[0], 20000, 2000));
+
+
+      // wait for remote bindings!
+      assertTrue(Wait.waitFor(() -> 1 == gatherBindingsOnT(0)[1], 20000, 2000));
+
+      assertTrue(Wait.waitFor(() -> 1 == gatherBindingsOnT(1)[1], 20000, 2000));
+
+      // produce a few every second with failover randomize=true to Topic T
+      producer.start();
+
+
+      assertTrue("All sent", Wait.waitFor(() -> toSend.get() == 0));
+
+      assertTrue("All received sub0", Wait.waitFor(() -> sub0.maxReceived == numMessages));
+
+      assertTrue("All received sub1", Wait.waitFor(() -> sub1.maxReceived == numMessages));
+
+
+      // with partition, none will be out of order
+      assertFalse(sub0.orderShot.get()  && sub1.orderShot.get());
+
+      sub0.consumerDone.set(true);
+      sub1.consumerDone.set(true);
+
+      stopServers(0, 1);
+   }
+
+   private void addBalancerWithClientIdConsistentHashMod(int max) {
+
+      for (int node = 0; node < max; node++) {
+         Configuration configuration = servers[node].getConfiguration();
+         BrokerBalancerConfiguration brokerBalancerConfiguration = new BrokerBalancerConfiguration().setName(BROKER_BALANCER_NAME);
+         brokerBalancerConfiguration.setTargetKey(TargetKey.CLIENT_ID).setLocalTargetFilter(TargetKeyResolver.DEFAULT_KEY_VALUE + "|" + node);
+         NamedPropertyConfiguration transformerConfig = new NamedPropertyConfiguration();
+         transformerConfig.setName(ConsistentHashModulo.NAME);
+         HashMap<String, String> properties = new HashMap<>();
+         properties.put(ConsistentHashModulo.MODULO, String.valueOf(max));
+         transformerConfig.setProperties(properties);
+         brokerBalancerConfiguration.setTransformerConfiguration(transformerConfig);
+
+         configuration.setBalancerConfigurations(Collections.singletonList(brokerBalancerConfiguration));
+
+         TransportConfiguration acceptor = getDefaultServerAcceptor(node);
+         acceptor.getParams().put("redirect-to", BROKER_BALANCER_NAME);
+      }
+   }
+
+   protected ConnectionFactory createFactory(String protocol, String clientID, String user, String password) throws Exception {
```

Comment:
is it possible to reuse the base method to avoid duplicating the common code?

Reply:
maybe, but it does not help with understanding. This use case needs failover and I thought it simpler to do here rather than add failover to the existing base. I guess it could go either way.